File: Math/BasicCalculator/BasicCalculator.py
# 224. Basic Calculator
# https://leetcode.com/problems/basic-calculator/


from typing import List


# 사칙연산 가능 계산기, 중위표현식을 후위표현식으로 바꾸는 정석?!
# 숫자는 바로 후위표현식에 넣음.
# 연산자는
#   연산자 스택에서 자기보다 우선순위가 높거나 같은 연산자를 빼서 후위표현식에 넣음,
#   자신을 연산자 스택에 넣음.
# 여는 괄호는 연산자 스택에 바로 넣음.
# 닫는 괄호는 여는 괄호를 만날 때까지 연산자 스택에서 빼서 후위표현식에 넣음.
# 끝나면 연산자 스택이 빌때까지 빼서 후위표현식에 넣음.
class Calculator:
    def calculate(self, s: str) -> int:
        s = s.replace(" ", "")
        self.op = {"(": 0, "*": 1, "/": 1, "+": 2, "-": 2}

        postExp = []
        opStack = []

        i = 0
        while i < len(s):
            if s[i].isdigit():
                temp = s[i]
                while i + 1 < len(s) and s[i + 1].isdigit():
                    i += 1
                    temp += s[i]
                postExp.append(temp)
            elif s[i] == ")":
                while opStack and opStack[-1] != "(":
                    postExp.append(opStack.pop())
                opStack.pop()
            elif s[i] in self.op:
                if (
                    opStack
                    and opStack[-1] != "("
                    and self.op[opStack[-1]] <= self.op[s[i]]
                ):
                    postExp.append(opStack.pop())
                opStack.append(s[i])

            i += 1

        while opStack:
            postExp.append(opStack.pop())

        print(postExp)
        return self.postfixCal(postExp)

    def postfixCal(self, exp: List[str]):
        stack = []
        for c in exp:
            if c.isdigit():
                stack.append(int(c))
            elif c in self.op:
                operand2 = stack.pop()
                operand1 = stack.pop() if stack else 0
                if c == "+":
                    stack.append(operand1 + operand2)
                elif c == "-":
                    stack.append(operand1 - operand2)
                elif c == "*":
                    stack.append(operand1 * operand2)
                elif c == "/":
                    stack.append(operand1 / operand2)

        return stack[0]


# O(n) : 후위표현식으로 변환 후 계산
# 으아 구현 구질구질하다.
class Solution:
    def calculate(self, s: str) -> int:
        s = s.replace(" ", "")
        self.op = set(["+", "-"])

        postExp = []
        opStack = []

        i = 0
        while i < len(s):
            if s[i].isdigit():
                temp = s[i]
                while i + 1 < len(s) and s[i + 1].isdigit():
                    i += 1
                    temp += s[i]
                postExp.append(temp)
                if opStack and opStack[-1] in self.op:
                    postExp.append(opStack.pop())

            elif s[i] in self.op:
                opStack.append(s[i])

            elif s[i] == "(":
                opStack.append(s[i])

            elif s[i] == ")":
                if opStack[-1] == "(":
                    opStack.pop()
                    if opStack:
                        postExp.append(opStack.pop())

            i += 1

        print(postExp)
        return self.postfixCal(postExp)

    def postfixCal(self, exp: List[str]):
        stack = []
        for c in exp:
            if c.isdigit():
                stack.append(int(c))
            elif c in self.op:
                operand2 = stack.pop()
                operand1 = stack.pop() if stack else 0
                if c == "+":
                    stack.append(operand1 + operand2)
                elif c == "-":
                    stack.append(operand1 - operand2)

        return stack[0]


def testSol(func, input, output):
    res = func(input)
    if res == output:
        print(f"O : {res}")
    else:
        print(f"X : {res}	excpet : {output}")


if __name__ == "__main__":
    sol = Solution()
    sol = Calculator()

    testSol(sol.calculate, "3210", 3210)
    testSol(sol.calculate, "(5)", 5)
    testSol(sol.calculate, "(1 + 1)", 2)
    testSol(sol.calculate, "1 + 1", 2)
    testSol(sol.calculate, " 2-1 + 2 ", 3)
    testSol(sol.calculate, "(1+(4+5+2)-3)+(6+8)", 23)
    testSol(sol.calculate, "(1+(4+54+2)-103)+(6+8)", -28)
    testSol(sol.calculate, "  -(  -2+3)", -1)

File: Math/BasicCalculator/BasicCalculator.java
package Math.BasicCalculator;
// 224. Basic Calculator
// https://leetcode.com/problems/basic-calculator/

/**
 * 구현이 참 더럽다.
 */

import java.util.Stack;


// O(n) 해답 : 부호만 스택에 저장 바로 바로 계산???
class Solution {
    public int calculate(String s) {
        int result = 0;
        int sign = 1;
        int num = 0;

        Stack<Integer> stack = new Stack<Integer>();
        stack.push(sign);

        for (char c : s.toCharArray()) {
            if (c >= '0' && c <= '9') {
                num = num * 10 + (c - '0');
            } else if (c == '+' || c == '-') {
                result += sign * num;
                sign = stack.peek() * (c == '+' ? 1 : -1);
                num = 0;
            } else if (c == '(') {
                stack.push(sign);
            } else if (c == ')') {
                stack.pop();
            }
        }

        result += sign * num;
        return result;
    }
}


// O(n) 스택에 넣으면서 계산.
class Solution1 {
    Stack<Character> opStack = new Stack<>();
    Stack<Integer> numStack = new Stack<>();

    public int calculate(String s) {
        s = s.replace(" ", "");

        this.opStack = new Stack<>();
        this.numStack = new Stack<>();

        int tempNum = 0;
        boolean start = true;
        for (int i = 0; i < s.length(); ++i) {
            char c = s.charAt(i);
            char op = ' ';
            if (start && c == '-') {
                numStack.push(0);
            }
            start = false;

            if (Character.isDigit(c)) {
                tempNum *= 10;
                tempNum += (c - '0');
                if (i + 1 == s.length() || !Character.isDigit(s.charAt(i + 1))) {
                    tempNum = _calculate(tempNum);
                    numStack.push(tempNum);
                    tempNum = 0;
                }
            } else if (c == '+' || c == '-') {
                opStack.push(c);
            } else if (c == '(') {
                opStack.push(c);
                start = true;
                continue;
            } else if (c == ')') {
                op = opStack.empty() ? ' ' : opStack.peek();
                while (op != '(') {
                    numStack.push(_calculate(numStack.pop()));
                    op = opStack.empty() ? ' ' : opStack.peek();
                    // System.out.println("op: " + this.opStack);
                    // System.out.println("num: " + this.numStack);
                }
                opStack.pop();
                while (!opStack.empty() && opStack.peek() != '(') {
                    numStack.push(_calculate(numStack.pop()));
                }
            }

            // System.out.println("op: " + this.opStack);
            // System.out.println("num: " + this.numStack);
        }

        return numStack.pop();
    }

    private int _calculate(int operand2) {
        if (this.opStack.empty() || this.numStack.empty() || this.opStack.peek() == '(') {
            return operand2;
        }
        int operand1 = this.numStack.pop();
        char op = this.opStack.pop();

        if (op == '+') {
            operand2 = operand1 + operand2;
        } else if (op == '-') {
            operand2 = operand1 - operand2;
        }
        return operand2;
    }
}


public class BasicCalculator {
    public static void main(String[] args) {
        testSol("", 0);
        testSol("1 + 1", 2);
        testSol("(0)", 0);
        testSol("(-1+3)", 2);
        testSol(" 2-1 + 2 ", 3);
        testSol("-1 + 1", 0);
        testSol("101 + 22", 123);
        testSol("(1+(4+5+2)-3)+(6+8)", 23);
        testSol(" -( -2+3)", -1);
        testSol("(7)-(0)+(4)", 11);
    }

    static void testSol(String input, int output) {
        Solution sol = new Solution();
        // todo : solution match
        int res = sol.calculate(input);
        if (res == output) {
            System.out.println("O : " + res);
        } else {
            System.out.println("x : " + res + "	expect : " + output);
        }
    }

}
